core/finality-grandpa/src/until_imported.rs

```diff
@@ -0,0 +1,565 @@
+			for &block_hash in self.pending.keys() {
+				if let Some(number) = self.status_check.block_number(block_hash)? {
+					known_keys.push((block_hash, number));
+				}
+			}
+
+			for (known_hash, canon_number) in known_keys {
+				if let Some(pending_messages) = self.pending.remove(&known_hash) {
+					let ready_messages = pending_messages.into_iter()
+						.filter_map(|m| m.wait_completed(canon_number));
+
+					self.ready.extend(ready_messages);
+				}
+			}
+		}
+
+		println!("Ready items? {}", self.ready.len());
+		if let Some(ready) = self.ready.pop_front() {
+			return Ok(Async::Ready(Some(ready)))
+		}
+
+		if self.import_notifications.is_done() && self.inner.is_done() {
+			Ok(Async::Ready(None))
+		} else {
+			Ok(Async::NotReady)
+		}
+	}
+}
+
+fn warn_authority_wrong_target<H: ::std::fmt::Display>(hash: H, id: AuthorityId) {
+	warn!(
+		target: "afg",
+		"Authority {:?} signed GRANDPA message with \
+		wrong block number for hash {}",
+		id,
+		hash,
+	);
+}
+
+impl<Block: BlockT> BlockUntilImported<Block> for SignedMessage<Block> {
+	type Blocked = Self;
+
+	fn schedule_wait<S, Wait, Ready>(
+		msg: Self::Blocked,
+		status_check: &S,
+		mut wait: Wait,
+		mut ready: Ready,
+	) -> Result<(), Error> where
+		S: BlockStatus<Block>,
+		Wait: FnMut(Block::Hash, Self),
+		Ready: FnMut(Self::Blocked),
+	{
+		let (&target_hash, target_number) = msg.target();
+
+		if let Some(number) = status_check.block_number(target_hash)? {
+			if number != target_number {
+				warn_authority_wrong_target(target_hash, msg.id);
+			} else {
+				ready(msg);
+			}
+		} else {
+			wait(target_hash, msg)
+		}
+
+		Ok(())
+	}
+
+	fn wait_completed(self, canon_number: NumberFor<Block>) -> Option<Self::Blocked> {
+		let (&target_hash, target_number) = self.target();
+		if canon_number != target_number {
+			warn_authority_wrong_target(target_hash, self.id);
+
+			None
+		} else {
+			Some(self)
+		}
+	}
+}
+
+/// Helper type definition for the stream which waits until vote targets for
+/// signed messages are imported.
+pub(crate) type UntilVoteTargetImported<Block, Status, I> = UntilImported<Block, Status, I, SignedMessage<Block>>;
+
+/// This blocks a commit message's import until all blocks
+/// referenced in its votes are known.
+///
+/// This is used for compact commits which have already been checked for
+/// structural soundness.
+pub(crate) struct BlockCommitMessage<Block: BlockT> {
+	inner: Arc<(AtomicUsize, Mutex<Option<(u64, CompactCommit<Block>)>>)>,
+	target_number: NumberFor<Block>,
+}
+
+impl<Block: BlockT> BlockUntilImported<Block> for BlockCommitMessage<Block> {
+	type Blocked = (u64, CompactCommit<Block>);
+
+	fn schedule_wait<S, Wait, Ready>(
+		input: Self::Blocked,
+		status_check: &S,
+		mut wait: Wait,
+		mut ready: Ready,
+	) -> Result<(), Error> where
+		S: BlockStatus<Block>,
+		Wait: FnMut(Block::Hash, Self),
+		Ready: FnMut(Self::Blocked),
+	{
+		use std::collections::hash_map::Entry;
+
+		enum KnownOrUnknown<N> {
```

Comment:
I didn't even know you could do that in rust. However, as this is only known within this scope, why have it generic over `N` and not just use `NumberFor<Block>` directly?

Though probably not a problem here, I am also wondering if that means the type is always bound to the execution scope and recreated on every call and whether that has a performance impact.

Reply:
Unfortunately, you can't bound the struct to local generic parameters. You have to make it generic itself.

Reply:
The type is still defined statically, it's just limited in namespace.